### PR TITLE
feat(tui): vim motions in prompt input

### LINKED ISSUE #1764: [FEATURE]: vim motions in input box
Please add the option of using vim keyboard shortcuts when writing the prompt (ClaudeCode has it, see attached)

Image

Image

### LINKED ISSUE #11111: [FEATURE]: VIM Keyboard Layout
### Feature hasn't been suggested before.

- [x] I have verified this feature I'm about to request hasn't been suggested before.

### Describe the enhancement you want to request

I would like to have the option to use the VIM keyboard layout to write the message. It’s really useful when working with long prompts. And this functionality already have the popular projects **Claude Code** and **Gemini Cli**.

Image

### PR DESCRIPTION
**Pull request by leohenon**
### What does this PR do?

Closes #1764
Closes #11111

Adds optional Vim motions to the prompt input. Enable with tui.vim: true or toggle from the menu. Disabled by default.

demo

Supported:
- Mode switching: `i I a A o O S`, `cc`, `cw`, `Esc`
- Motions: `h j k l`, `w b e`, `W B E`, `0 ^ $`
- Deletes: `x`, `dd`, `dw`
- Session navigation: `gg/G`
- Scrolling: `Ctrl+e/y/d/u/f/b`
- `Enter` in normal mode submits

Implementation: Vim state isolated in `/vim` module. Handler owns mode + pending operator state. Mode resets to insert after submit. Visual mode intentionally excluded.

I saw thdxr's concern in #1764 about partial Vim feeling incomplete. This PR covers only the core motions needed for prompt editing and is intentionally kept small to make review manageable. I have a fuller implementation with yank/paste, visual mode and other motions that I can split into follow-up PRs if this direction looks good or keep it at this core subset as it seems many people are satisfied with a minimal vim mode.

### How did you verify your code works?

I use the fuller implementation daily. Reimplemented the core subset for this PR and verified manually (editing, scrolling, session switching, edge cases with pending operators, empty buffers, multiline). Added targeted tests for mode transitions, motions, deletes, pending state, and submit reset. 

Not knowing when this will be reviewed - If you would like to use this now with a more complete vim mode, use [opencode-vim](https://github.com/leohenon/opencode).

### CODE REVIEW

File: packages/opencode/src/cli/cmd/tui/component/vim/vim-motions.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  return pos
+}
+
+function wordEnd(text: string, offset: number, big: boolean) {
+  const match = big ? isBigWord : isWord
+  let pos = offset
+  if (pos < text.length) pos++
+  while (pos < text.length && !match(text[pos])) pos++
+  while (pos < text.length && match(text[pos])) pos++
+  if (pos > offset + 1) pos--
```
[the remark this answers is not in the document]
Comment (leohenon) [reply packages/opencode/src/cli/cmd/tui/component/vim/vim-motions.ts]: fixed

File: packages/opencode/src/cli/cmd/tui/component/vim/vim-state.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { createEffect, createMemo, createSignal, type Accessor } from "solid-js"
+
+export type VimMode = "normal" | "insert"
+export type VimPending = "" | "d" | "g"
+
+export function createVimState(input: {
+  enabled: Accessor<boolean>
+  active: Accessor<boolean>
+  initial?: Accessor<VimMode | undefined>
+}) {
```
[the remark this answers is not in the document]
Comment (leohenon) [reply packages/opencode/src/cli/cmd/tui/component/vim/vim-state.ts]: fixed

**Comment by broskees** [comment]
This would be so good.

**Comment by kaich** [comment]
Good, This is exactly what I wanted.

**Comment by s-yakubovskiy** [comment]
Waiting for this as well <3

**Comment by Alimedhat000** [comment]
Exactly what i wanted

**Comment by broskees** [comment]
This is big!

**Comment by hamidzr** [comment]
thanks for the work, this is great and I need it! Not knowing how long it'll take the team to get to merging this I put up [a branch on my fork](https://github.com/hamidzr/opencode/tree/az) that will track the latest stable release + this + some additions I had around vim motion + a simple install script. I'll be using it and I thought I'd share in case you also can't wait to use it!

**Comment by ppiwko** [comment]
@leohenon, could you please rebase this on the latest OpenCode version? Alternatively, would you consider releasing your own Vim-based version of OpenCode? I’m currently using your fork as my primary coding agent, and it’s been great! 👍

**Comment by leohenon** [comment]
> @leohenon, could you please rebase this on the latest OpenCode version? Alternatively, would you consider releasing your own Vim-based version of OpenCode? I’m currently using your fork as my primary coding agent, and it’s been great! 👍

Released a vim build that tracks upstream - [ocv](https://github.com/leohenon/opencode).

**Comment by ppiwko** [comment]
> > @leohenon, could you please rebase this on the latest OpenCode version? Alternatively, would you consider releasing your own Vim-based version of OpenCode? I’m currently using your fork as my primary coding agent, and it’s been great! 👍
> 
> Released a vim build that tracks upstream - [ocvim](https://github.com/leohenon/opencode).

Could you please enable Issues for the OCV project? I’d like to submit a feature request for a Vim-style 'Shift+J' line-join shortcut.

**Comment by mahersaba** [comment]
Hi, I really can't wait for this to get in. Do you need help?
Also, can we add a /vim to toggle the vim mode so it's automatically in <insert> mode by default on every trun?

**Comment by ppiwko** [comment]
The [ocv fork from leohenon](https://github.com/leohenon/opencode) is fantastic! It currently supports long list of Vim motions, including:

h j k l w b e W B E 0 ^ _ $ gg G i I a A o O x dd dw cc cw S J yy yw p P v V f F t T ; , Ctrl+e Ctrl+y Ctrl+d Ctrl+u Ctrl+f Ctrl+b

It's the best terminal AI agent for Vim users! 😍

**Comment by kpetroff** [comment]
vim mode in opencode would be big, can't wait this to happen

**Comment by leohenon** [comment]
@mahersaba Vim mode is already persistent via local KV. No slash command, you toggle it from the command palette.

**Comment by leohenon** [comment]
@kpetroff You can install it from my fork in the meantime, it has more motions implemented and syncs with latest opencode releases - https://github.com/leohenon/opencode

**Comment by Hekmoo** [comment]
Maybe this might be better suited as a plugin?

**Comment by gotgenes** [comment]
> Maybe this might be better suited as a plugin?

I would agree except the plugin ecosystem has not been getting much love from OpenCode maintainers, e.g., #12042 lingers…

**Comment by Hekmoo** [comment]
It looks like it will get some love here soon according to dax.

**Comment by leohenon** [comment]
> Maybe this might be better suited as a plugin?

I would agree, but unfortunately opencode doesn’t expose a prompt keydown hook yet.

**Comment by aizigao** [comment]
I use your fork as my primary open-code now

**Comment by ccolorado** [comment]
So only the ut and merge conflicts preventing this to be added ?

**Comment by VimelWei** [comment]
I tested OpenCode Vim for a day and encountered compatibility issues with Chinese and other multi-byte languages:

 - After navigating with word motions like w or b, the cursor disappears when it lands on Chinese characters.
- The $ motion does not jump to the actual end of the line when multi-byte characters are present.

This appears related to the known limitation that Unicode word boundaries are not supported.

**Comment by leohenon** [comment]
Given the recent plugin improvements, I’m not sure this needs to land in core anymore. This PR can still serve as a reference implementation.

I don’t plan to expand the scope further unless maintainers want this. In particular, full unicode word-boundary / display-width support.

If the preferred direction is plugin only, I’m fine closing this in favor of that effort.

[opencode-vim](https://github.com/leohenon/opencode-vim) will continue to be maintained for features that don’t seem plugin-friendly yet, mainly visual and copy mode, though CJK support is still a limitation for now.

**Comment by rsdrahat** [comment]
+1 for vim motions in the prompt input &mdash; this is the feature that keeps me on Claude Code when I test-drive opencode. Thanks for the implementation in this PR and for maintaining [ocv](https://github.com/leohenon/opencode-vim) as a workable interim solution. Curious if the maintainers have a preference between core integration vs. a new plugin API event (e.g. `tui.input.before`) that would let vim mode ship as a plugin. Happy to help push either path forward.

**Comment by kriscoleman** [comment]
I just shared with my team that I wish I could use a vim mode instead of mouse awareness. 
A teammate then pointed me to this PR. 

This would be lovely. 🚀

**Comment by m2cci-bouzentm** [comment]
Thanks for this! Hope it gets merged soon.

**Comment by jonnjonnjo** [comment]
hopefully this merged soon 😄

**Comment by fsilly** [comment]
> Given the recent plugin improvements, I’m not sure this needs to land in core anymore. This PR can still serve as a reference implementation.
> 
> I don’t plan to expand the scope further unless maintainers want this. In particular, full unicode word-boundary / display-width support.
> 
> If the preferred direction is plugin only, I’m fine closing this in favor of that effort.
> 
> [opencode-vim](https://github.com/leohenon/opencode-vim) will continue to be maintained for features that don’t seem plugin-friendly yet, mainly visual and copy mode, though CJK support is still a limitation for now.

@leohenon Hello, I couldnt find any [plugins](https://opencode.ai/docs/ecosystem#plugins) allowing vim controls inside opencode. I didnt understand what you meant.

Thank you for reading.

**Comment by leohenon** [comment]
@fsilly I’m not aware of any working plugin.

I meant that the basic vim motions in this PR should now be possible via
the new plugin system, whereas they weren’t before.

I’m not planning to build a plugin or keep this PR up to date since I’m focusing on [ocv](https://github.com/leohenon/opencode-vim), but this can be a reference if anyone wants to build one.

**Comment by claytonweaver** [comment]
Get this merged!!!

**Comment by leohenon** [comment]
Update: I ended up building the plugin [opencode-vim-plugin](https://github.com/leohenon/opencode-vim-plugin):

 ```bash
   opencode plugin @leohenon/opencode-vim-plugin --global
 ```

 It covers everything in this PR and more.

 With the plugin available I'm closing this,  it covers everything here without core changes. [opencode-vim](https://github.com/leohenon/opencode-vim) remains for what plugins can't reach yet.

**Comment by augustocdias** [comment]
@leohenon what your version has that the plugin doesn't? I'm considering if I should use the plugin or your fork

**Comment by leohenon** [comment]
@augustocdias The fork adds copy mode, including subagent and tool navigation. The plugin has essentially everything else for prompt editing, apart from a few configuration options that may be added later. If you only need vim prompt editing, then the plugin is sufficient. 

[Discussion #183](https://github.com/leohenon/opencode-vim/discussions/183)
